pkg/controller/daemon/daemoncontroller_test.go

```diff
@@ -533,6 +547,27 @@ func TestInsufficientCapacityNodeDaemonDoesNotUnscheduleRunningPod(t *testing.T)
 	}
 }
 
+// DaemonSets should only place onto nodes with sufficient free resource and matched node selector
+func TestInsufficientCapacityNodeSufficientCapacityWithNodeLabelDaemonLaunchPod(t *testing.T) {
+	podSpec := resourcePodSpecWithoutNodeName("50M", "75m")
+	ds := newDaemonSet("foo")
+	ds.Spec.Template.Spec = podSpec
+	ds.Spec.Template.Spec.NodeSelector = simpleNodeLabel
+	manager, podControl, _ := newTestController(ds)
+	node1 := newNode("not-enough-resource", nil)
+	node1.Status.Allocatable = allocatableResources("10M", "20m")
+	node2 := newNode("enough-resource", simpleNodeLabel)
+	node2.Status.Allocatable = allocatableResources("100M", "200m")
+	manager.nodeStore.Add(node1)
+	manager.nodeStore.Add(node2)
+	manager.dsStore.Add(ds)
+	syncAndValidateDaemonSets(t, manager, ds, podControl, 1, 0)
+	// we do not expect any event for insufficient free resource
+	if len(manager.fakeRecorder.Events) != 0 {
```

Comment:
I will try to move the `event number` check to `syncAndValidateDaemonSets` in https://github.com/kubernetes/kubernetes/issues/46629 and then if the `event number` is greater than 0, check `event reason` after `syncAndValidateDaemonSets`, will post a PR once this get merged.